man/nextxrec.3
.TH NEXTXREC 3 "10 April 1996"
.FB
.SH NAME
nextxrec, prevxrec \- load next/previous indexed record
.SH SYNOPSIS
#include <firstbase/fb.h>
.sp 1
fb_nextxrec(db)
.br
fb_database *db;
.sp 1
fb_prevxrec(db)
.br
fb_database *db;
.PP
.SH DESCRIPTION
As explained in the \fIdatabase\fP(4) manual page,
a database structure contains
all of the elements needed for an entire database, including an array of
field structures, all file names and descriptors, and index information.
Altogether, the array of field structures can hold one complete record.
.PP
\fIFb_nextxrec\fP attempts to load the next indexed record into
database \fIdb\fP.
If a record has not already been loaded using \fIfb_getxrec\fP(3) or a
similar index mechanism, then the initial call (and only the initial call)
to \fIfb_nextxrec\fP will load the 
the first indexed record (ala \fIfirstxrec\fP(3)).
To repeat a loop through all indexed records, use \fIfirstxrec\fP(3)
followed by \fInextxrec\fP, or use the \fIforxeach\fP(3) mechanism.
.PP
\fIFb_prevxrec\fP attempts to load the previous indexed record into database
\fIdb\fP.
If one record has not already been loaded, nothing is done.
.PP
Once a record number is retrieved from the index, \fIgetrec\fP(3) is used
by these routines to
actually load the record into the database structures.
.PP
When using a \fIFirstBase\fP flat index, 
the physical position number of the index entry, the index record number,
is stored in the database structure \fIdb->bsrec\fP.
If \fIdb->bsrec\fP is greater
than \fIdb->bsend\fP, the returned record's key is stored in the overflow area
of the index, otherwise it is stored in the binary searchable area.
.PP
However, when using a \fIFirstBase\fP Btree+ index, the \fIdb->bsrec\fP value
is actually the record number (not the index record number).
.SH CLIENT/SERVER
The \fIfb_nextxrec\fP and \fIfb_prevxrec\fP routines will transparently
call the \fIfb_nextxrec_clnt\fP and \fIfb_prevxrec_clnt\fP
mechanisms when the global \fIFirstBase\fP variable \fIcdb_use_rpc\fP
is set to one. See \fIfb_clnt_create\fP(3) for more details.
.SH SEE ALSO
getxrec(3), getrec(3), firstxrec(3), opendb(3), database(4), index(4).
.SH DIAGNOSTICS
These routines will return FB_AOK on a successful call, and FB_ERROR when
the record could not be found or loaded.
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/librec/nextxrec.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: nextxrec.c,v 9.0 2001/01/09 02:57:00 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Nextxrec_sid[] = "@(#) $Id: nextxrec.c,v 9.0 2001/01/09 02:57:00 john Exp $";
#endif

#include <fb.h>
#include <fb_ext.h>

extern short int cdb_use_rpc;
extern short int cdb_secure;
extern long fb_key_eval(fb_bseq *bs);

/*
 *  nextxrec - force the next indexed record to be loaded.
 */

   fb_nextxrec(dp)
      fb_database *dp;
      
      {
         long rec, save_recno;
	 int st = FB_ERROR, save_curkey;
         fb_bseq *bs;
	 
#if RPC
         if (cdb_use_rpc)
            return(fb_nextxrec_clnt(dp));
#endif /* RPC */
         if (dp->b_tree){				/* btree path */
            bs = dp->b_seq;
            if (bs->bs_recno == 0){
               bs->bs_next = 0;
               bs->bs_curkey = 0;
               fb_get_seq_head(bs);
               if (fb_seq_getrec(bs->bs_head, bs) == FB_ERROR)
                  return(FB_ERROR);
               }
            save_recno = bs->bs_recno;
            save_curkey = bs->bs_curkey;
            /*
             * the current node has to be re-read in case a key was deleted
             * by some other process.
             */
            if (bs->bs_curkey >= 1 && bs->bs_curkey <= 3)
               fb_seq_getrec(bs->bs_recno, bs);
            for (;;){
               if (bs->bs_curkey >= 3){
                  if (bs->bs_next == 0)
                     break;
                  if (fb_seq_getrec(bs->bs_next, bs) != FB_AOK){
                     st = FB_ERROR;
                     break;
                     }
                  bs->bs_curkey = 0;
                  }
               bs->bs_curkey++;
               rec = fb_key_eval(bs);
               if (rec > 0L && rec <= dp->reccnt){	/* ignore rec of 0 */
                  if (fb_getrec(rec, dp) == FB_ERROR)
                     fb_xerror(FB_FATAL_GETREC, dp->dbase, (char *) &rec);
                  if (cdb_secure && fb_record_permission(dp, READ) == FB_ERROR)
                     continue;
                  st = FB_AOK;
                  dp->bsrec = rec;
                  break;
                  }
               else if (rec == 0)
                  bs->bs_curkey = 4;
               else if (rec < -1){
                  st = FB_ERROR;
                  break;
                  }
               }
            /* if not found, restore saved location */
            if (st == FB_ERROR){
               bs->bs_recno = save_recno;
               bs->bs_curkey = save_curkey;
               }
            }
         else {						/* standard FB index */
            while (read(dp->ifd, dp->irec, (unsigned) dp->irecsiz) ==
                  dp->irecsiz){
               dp->bsrec += 1L;
               rec = atol((char *) (dp->irec + dp->irecsiz - 11));
               if (rec > 0L && rec <= dp->reccnt){	/* ignore rec of 0 */
                  if (fb_getrec(rec, dp) == FB_ERROR)
                     fb_xerror(FB_FATAL_GETREC, dp->dbase, (char *) &rec);
                  st = FB_AOK;
                  if (cdb_secure && fb_record_permission(dp, READ) == FB_ERROR)
                     st = FB_ERROR;
                  break;
                  }
               }
            }
	  return(st);
      }
